datatype: ensure is_contig is consistent with yaksa

hzhou opened the pull request:
## Pull Request Description

Datatype currently sets `is_contig` baseds on its own logic, which may
end up inconsistent with how yaksa thinks. It is a separate issue
whether yaksa is correct or mpich is correct, at least we need ensure
consistency. This commit asks for `num_contig_blocks` with multiple
count to deduce `is_contig`.


<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->
## NOTE
* [x] There is a pending issue in yaksa https://github.com/pmodels/yaksa/issues/111 needs to be fixed before this PR can be merged.

[skip warnings]
## Background
Related yaksa issue: https://github.com/pmodels/yaksa/issues/109

This inconsistency is currently causing a failure in `rma/getfence1 -type=MPI_INT -count=262144 -seed=793 -testsize=16`, where MPICH thinks a datatype is contiguous, thus calls `MPIDI_OFI_nopack_putget`, but yaksa gives 262144 very low density iov, resulting in slow execution. Currently this only fails when multiple vci is enabled. It appears with single vci, `262144` `fi_readmsg` is still quick enough, but when multiple vci is enabled (even when RMA only uses vci 0), the same `fi_readmsg` takes hundreds of retry for at least the `sockets` provider. This behavior is being investigated separately.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ofi
env: MPIR_CVAR_CH4_NUM_VCIS=4

pavanbalaji commented:
Why are we setting the type information in MPICH and then overwriting it in typerep?  That sounds redundant.  We should always get it in typerep and not recalculate it in MPICH.

hzhou commented:
> 
> 
> Why are we setting the type information in MPICH and then overwriting it in typerep? That sounds redundant. We should always get it in typerep and not recalculate it in MPICH.

I agree with the sound. 😀.  I was looking for shortest distance here. 

First, since the dataloop code doesn't populate these fields, we probably need copy/move these code to the dataloop functions.

Second, does yaksa expose all these informations? I am using a hack to get the `is_contig` filed here simply because yaksa doesn't expose that.

pavanbalaji commented:
I agree that, if the change becomes too intrusive, that won't be a good solution.  Do you have an estimate on how much code churn it would cause to move the querying of this information to typerep?  Maybe we decide based on that?

hzhou commented:
> 
> 
> I agree that, if the change becomes too intrusive, that won't be a good solution. Do you have an estimate on how much code churn it would cause to move the querying of this information to typerep? Maybe we decide based on that?

There will be 10~30 lines in every MPI- datatype creation functions that need be moved inside corresponding `typerep_dataloop_xxx`. Then yaksa need expose those information (`lb, true_lb, sz, extent, is_contig, alignsize`, some is necessary, some maybe omitted) to mpich. The shortest way maybe ask yaksa to make its type structure public; otherwise, you need have public API function query them. Then there probably will be additional several lines in each `typerep_yaksa_xxx` functions to fill/copy the fields.

The blocker is yaksa need expose the fields. The changes probably around 300 to 400 lines, but not difficult.

pavanbalaji commented:
I'm not too concerned about the changes in typerep, as that's a fairly independent and mostly untouched-by-others component.  I don't think we need any functionality changes in yaksa, except for the bug-fix pointed out in https://github.com/pmodels/yaksa/issues/109.

But the changes in the MPI datatype creation functions is the only part that's a concern.  If that's a small-enough change, I'd say go for it.  If you think that part is too much churn, then we can reconsider this shortcut.

hzhou commented:
> I don't think we need any functionality changes in yaksa.

If yaksa doesn't expose those fields, then  we are talking about *duplicating* the current `MPI-level` code inside both `typerep_dataloop_xxx` and `typerep_yaksa_xxx`. For  yaksa path, we are still deriving those fields separately, once in mpich and once inside yaksa, no way to ensure consistency. So compare to that, I think I would much prefer current shortcut.

pavanbalaji commented:
I meant that yaksa already exposes them, but not as a separate structure.  You'd need to make the appropriate yaksa calls to get them.  Am I missing something?  Is there a piece of information that MPICH computes that yaksa doesn't expose?

hzhou commented:
> I meant that yaksa already exposes them, but not as a separate structure. You'd need to make the appropriate yaksa calls to get them. Am I missing something? Is there a piece of information that MPICH computes that yaksa doesn't expose?

Like I am the one missing something. What is the yaksa functions to get `lb`, `ub`, `size`, `is_contig`?

pavanbalaji commented:
You can use `yaksa_type_get_size` to get the size; `yaksa_type_get_true_extent` to get the true lb, ub, extent; `yaksa_type_get_extent` to get the lb, ub, extent.  For `is_contig`, there's no direct function, but you can use `yaksa_iov_len` with count 2 (similar to what your current patch does) to get that information.

hzhou commented:
Okay, sounds good. Do you want give a go-ahead? It is a few hundred lines of churn, but I don't think anyone is touching that right now.

pavanbalaji commented:
> Okay, sounds good. Do you want give a go-ahead? It is a few hundred lines of churn, but I don't think anyone is touching that right now.

Can you try to keep the changes to the MPI datatype creation code minimal?  Intrusive changes to typerep are fine.

hzhou commented:
> > Okay, sounds good. Do you want give a go-ahead? It is a few hundred lines of churn, but I don't think anyone is touching that right now.
> 
> Can you try to keep the changes to the MPI datatype creation code minimal? Intrusive changes to typerep are fine.

The current code calculates those fileds inside the MPI datatype creation functions, which are the ones you want to remove, right?

pavanbalaji commented:
> The current code calculates those fileds inside the MPI datatype creation functions, which are the ones you want to remove, right?

Yes, unless you see a way to do this without that churn either.

hzhou commented:
> Yes, unless you see a way to do this without that churn either.

You cannot remove the code without touching it 😄

hzhou commented:
@pavanbalaji The comment in `yaksa.h.in` says `yaksa_type_get_true_extent`  and `yaksa_type_get_extent` are
```
[in] type
[in] lb
[out] extent
```

... should be `[out] lb`, right?

pavanbalaji commented:
Oh, yes.  Good catch.  Would you be able to submit a PR to fix it?

hzhou commented:
@pavanbalaji 
* updated first commit to let yaksa update `lb`, `ub`, `size`, etc.
* added two commits to move the redundant code from MPI-type-creation into `typerep_dataloop_create.c`. Just `resized` and `struct` so far. More to come. Let me know do you still want go-ahead.
* Can yaksa expose `builtin_element_size` and `n_builtin_elements` -- needed by `MPIR_Type_get_elements` and it is a mess if we still need derive that in MPI-type-creation, just look at the `struct` commit.

pavanbalaji commented:
Yeah, I think the removal is clean enough to use for other types too.

Btw, `alignsize` is a dataloop-specific piece of information.  It does not need to be in the `MPIR_Datatype` structure as it would not be used by MPICH -- it can go into some dataloop internal data structure.  Yaksa has an equivalent piece of information that it maintains, but it doesn't expose it.

hzhou commented:
> Btw, `alignsize` is a dataloop-specific piece of information.

OK.

And don't forget this question: 

>> Can yaksa expose builtin_element_size and n_builtin_elements?

The logic in getting that is still quite messy with some types (look at the `struct`, currently incomplete). If yaksa doesn't expose this, the much of the code such as in `struct` still need be retained in MPI-layer.

hzhou commented:
Tagging @pavanbalaji for the question in https://github.com/pmodels/mpich/pull/4758#issuecomment-682218933

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most

hzhou commented:
@pavanbalaji There are test failures with yaksa `lb` and `ub`: https://jenkins-pmrs.cels.anl.gov/view/mpich-review/job/mpich-review-ch4-ofi/1597/testReport/

It appears that `yaksa` doesn't get the evil datatype correctly:
```
[3] contig (count=3)
[2] resized (lb=6, extent=-9)
[1] contig (count=4)
base: MPI_BYTE
```
The supposed type map: `{ (LB,6),4*(BYTE,0),(UB,-3),(LB,-3),4*(BYTE,-9),(UB,-12),(LB,-12),4*(BYTE,-18),(UB,-21) }`

Error message:
```
##   lb of type = 22; should be -12
##   extent of type = -27; should be 9
##   true_lb of type = 0; should be -18
##   true extent of type = -14; should be 22
## found 4 errors in negextent contig test
```

pavanbalaji commented:
LOL.  It's quite appropriately named as the "evil datatype".  I'll take a look at it.

With respect to the `builtin_element_size` and `n_builtin_elements`, I'm curious as to how that would even work for struct types.  I'll need to dig deeper into the code to understand their semantics a bit better before I can respond to whether we can/should add them to yaksa.

hzhou commented:
> With respect to the `builtin_element_size` and `n_builtin_elements`, I'm curious as to how that would even work for struct types. I'll need to dig deeper into the code to understand their semantics a bit better before I can respond to whether we can/should add them to yaksa.

I duplicated the element deduction code into `typerep_yaksa_create.c` for now: https://github.com/pmodels/mpich/pull/4758/commits/f893126f84408bcf92a2de95746f6b4c41b443f8. With `struct`, it is opportunistic -- setting correct element if it is uniform, otherwise, `-1`. Apparently, `MPI_get_element` behave gracefully with `-1`.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

pavanbalaji commented on src/mpi/datatype/typerep/src/typerep_yaksa_create.c:
```diff
@@ -7,6 +7,33 @@
+        MPIR_ERR_CHKANDJUMP(rc, mpi_errno, MPI_ERR_INTERN, "**yaksa"); \
+        newtype->true_ub = newtype->true_lb + true_extent; \
+    } while (0)
+
```
Why is this a macro?  It's not performance critical, and it doesn't require any state from the calling function.  In general, we have been moving away from macros in MPICH towards either functions or static inlines.

hzhou replied:
I'll make it into a function

pavanbalaji commented on src/mpi/datatype/typerep/src/typerep_yaksa_create.c:
```diff
@@ -49,6 +49,19 @@ int MPIR_Typerep_create_vector(int count, int blocklength, int stride, MPI_Datat
+        newtype->n_builtin_elements = count * blocklength * old_dtp->n_builtin_elements;
+        newtype->builtin_element_size = old_dtp->builtin_element_size;
+        newtype->basic_type = old_dtp->basic_type;
+    }
```
Shouldn't this go into the newly created `YAKSA_UPDATE_NEWTYPE` function/macro?

hzhou replied:
This is the thing with function vs macro. Once we make it into a function, it becomes an independent compilation entity and it needs better, more complete, and less flexible abstraction. For example, once we move the above-mentioned code into the function, the function will host code from all different datatypes into a general context. In another view, we are pulling code from their own context where is narrow and simple into a general context where it requires an understanding of all datatype contexts. Function, in my view, incurs more accidental complexity than macros.

But here, I think we can manage. So I will make the change.

pavanbalaji commented on src/mpi/datatype/type_dup.c:
```diff
@@ -62,12 +62,12 @@ int MPIR_Type_dup(MPI_Datatype oldtype, MPI_Datatype * newtype)
         }
-
-        mpi_errno = MPIR_Typerep_create_dup(oldtype, new_dtp);
-        MPIR_ERR_CHECK(mpi_errno);
```
I believe this is only a dataloop weirdness.  Yaksa doesn't require this special commit after dup.

hzhou replied:
Yeah.

pavanbalaji replied:
So, perhaps this can be moved to dataloop?

pavanbalaji reviewed (changes requested):
This PR looks mostly great.  I have left some minor comments.

hzhou commented:
removing the extra type_commit in MPIR_Type_dup and see if we are missing anything:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
Updated the code that
* use function instead of macro
* move element update code into the internal function. The first two commits are now squashed together as the split was merely an artifact during development.
* moved the `MPID_Type_commit` fix up to group it together with the commit that update MPI_Type_dup.
* rebased to main

The commit remains as the `MPID_Type_commit` in `MPI_Type_dup` is needed not because of dataloop. It is because `MPI_Type_due` doesn't require user to run `MPI_Type_commit` if the old type is already committed, but `ucx` need to run the commit hook regardless in order to register the datatype with `ucp`.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
@pavanbalaji Ready for your review again

pavanbalaji commented:
> `MPI_Type_due` doesn't require user to run `MPI_Type_commit` if the old type is already committed, but `ucx` need to run the commit hook regardless in order to register the datatype with `ucp`.

Does the MPI standard say that the user does not need to run `MPI_Type_commit` after `dup`?  My impression was that a commit was required after a dup.

hzhou commented:
> > `MPI_Type_due` doesn't require user to run `MPI_Type_commit` if the old type is already committed, but `ucx` need to run the commit hook regardless in order to register the datatype with `ucp`.
> 
> Does the MPI standard say that the user does not need to run `MPI_Type_commit` after `dup`? My impression was that a commit was required after a dup.

Here is quote from mpi standards (on `MPI_Type_dup`):
```
Returns in newtype a new datatype with exactly the same properties as oldtype and any copied cached information, see Section 6.7.4. The new datatype has identical upper bound and lower bound and yields the same net result when fully decoded with the functions in Section 4.1.13. The newtype has the same committed state as the old oldtype.
```
